src/ucp/core/ucp_worker.c

```diff
@@ -385,6 +415,87 @@ void ucp_worker_progress(ucp_worker_h worker)
     ucs_assert(--worker->inprogress == 0);
 }
 
+ucs_status_t ucp_worker_get_efd(ucp_worker_h worker, int *fd)
+{
+    unsigned i;
+    ucs_status_t status;
+    ucp_context_h context = worker->context;
+    struct epoll_event event;
+    event.events = EPOLLIN | EPOLLET;
+
+    if (worker->wakeup_efd != -1) {
+        *fd = worker->wakeup_efd;
+        return UCS_OK;
+    }
+
+    *fd = epoll_create(context->num_tls);
+    if (*fd == -1) {
+        return UCS_ERR_IO_ERROR;
+    }
+
+    for (i = 0; i < context->num_tls; i++) {
+        if (worker->iface_wakeups[i]) {
+            status = uct_wakeup_efd_get(worker->iface_wakeups[i],
+                    &event.data.fd);
+            if (status != UCS_OK) {
+                close(*fd);
+                return status;
+            }
+
+            if (epoll_ctl(*fd, EPOLL_CTL_ADD, event.data.fd, &event))
+            {
+                close(*fd);
+                return status;
+            }
+        }
+    }
+
+    return UCS_OK;
+}
+
+ucs_status_t ucp_worker_arm(ucp_worker_h worker)
+{
+    ucs_status_t status;
+    ucp_rsc_index_t tl_id;
+    ucp_context_h context = worker->context;
+
+    for (tl_id = 0; tl_id < context->num_tls; ++tl_id) {
+        status = uct_wakeup_efd_arm(worker->iface_wakeups[tl_id]);
+        if (status != UCS_OK) {
+            return status;
+        }
+    }
+
+    return UCS_OK;
+}
+
+ucs_status_t ucp_worker_wait(ucp_worker_h worker)
+{
+    int epoll_fd;
+    struct epoll_event *events;
+    ucs_status_t status = ucp_worker_get_efd(worker, &epoll_fd);
+
+    if (status == UCS_OK) {
+        ucp_context_h context = worker->context;
+        events = alloca(context->num_tls * sizeof(*events));
```

Comment:
I think I need it, otherwise epoll should return "EFAULT - The memory area pointed to by events is not accessible with write permissions."